libs/ngxtension/reduce-array/src/reduce-array.ts

```diff
@@ -0,0 +1,33 @@
+import { Observable } from 'rxjs';
+import { map } from 'rxjs/operators';
+
+export function reduceArray<T>(
+	reduceFn: (acc: T, item: T, index: number) => T
+): (source: Observable<T[]>) => Observable<T>;
+
+export function reduceArray<T, R = T>(
+	reduceFn: (acc: R, item: T, index: number) => R,
+	initialValue?: R
+): (source: Observable<T[]>) => Observable<R>;
+
+export function reduceArray<T>(
+	reduceFn: (acc: any, item: T, index: number) => any,
+	initialValue?: any
+): (source: Observable<T[]>) => Observable<any> {
+	return map((array: T[]) => {
```

Comment:
nit/suggestion: the `T` for the implementation with overloads is redundant. Clean them all up with `any`. However for this case, that might be too many `any`. So here's my suggestion:

```ts
type AnyArray = Array<any>;
type ReduceParameters = Parameters<AnyArray['reduce']>;

export function reduceArray<T>(
	reduceFn: (acc: T, item: T, index: number) => T
): (source: Observable<T[]>) => Observable<T>;

export function reduceArray<T, R = T>(
	reduceFn: (acc: R, item: T, index: number) => R,
	initialValue: R
): (source: Observable<T[]>) => Observable<R>;

export function reduceArray(
	reduceFn: ReduceParameters[0],
	initialValue?: ReduceParameters[1]
) {
	return map((array: AnyArray) => {
		// call reduce function with initialValue
		if (initialValue !== undefined) {
			return array.reduce(reduceFn, initialValue);
		}
		// no initialValue

		// Javascript throws error if array is empty: [].reduce((acc,n) => acc +n)
		// avoid errors and return undefined
		if (!array.length) {
			return undefined;
		}
		// if array is not empty, call the reduceFn without initialValue
		return array.reduce(reduceFn);
	});
}
```

Reply:
I think any is easier to understand, and not a problem inside overloaded function. But you way is more the TS way but feel harder to fully get the why 😅

Reply:
what about:

`type Reduce = Array<any>['reduce'];`

then:
```ts
export function reduceArray(
	reduceFn: Parameters<Reduce>[0],
	initialValue?: Parameters<Reduce>[1]
) {
	return map((array: Array<any>) => {
```

?

Reply:
I'm ok with that

Reply:
Review now…